ethcore/sync/src/chain/syncpacketid.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity Ethereum.
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+//! When sending packets over p2p we specify both which subprotocol
+//! to use and what kind of packet we are sending (through a packet id).
+//! Likewise when receiving packets from other peers we decode the
+//! subprotocol and the packet id. This module helps coupling both
+//! pieces of information together and provides an easy mechanism
+//! to convert to/from the packet id values transmitted over the
+//! wire.
+
+use api::{ETH_PROTOCOL, WARP_SYNC_PROTOCOL_ID};
+use network::{PacketId, ProtocolId};
+
+/// An enum that defines all known packet ids in the context of
+/// synchronization and provides a mechanism to convert from
+/// packet ids (of type PacketId or u8) directly read from the network
+/// to enum variants. This implicitly provides a mechanism to
+/// check whether a given packet id is known, and to prevent
+/// packet id clashes when defining new ids.
+enum_from_primitive! {
```

Comment:
I'm not sure if the macro is worth additional (seems like not very well maintained) dependency.
Can we implement `from_usize/from_u8` manually?

Alternatively could we use constants instead of enum?
```
const STATUS_PACKET = SyncPacketId(0x00, ETH_PROTOCOL);
```

Reply:
I considered using constants and found one argument against them: we have to manually check that no two constants repeat the same value, or they don't end up claiming the same packet_id, that's why I thought an enum would be a good idea.

We can indeed implement the conversion manually, though I found it repetitious: adding one new Id would mean adding it to the enum and to the conversion methods: three places to touch.

Agreed, the package seems a bit outdated, even if it has a large number of downloads. If we drop the dependency, for cleanliness I would go for constans.

Reply:
Well now we need to maintain a map of `variant -> protocol_id` manually.
Perhaps we can have a custom (proc) macro for this to cater all cases?

Reply:
Yeah, that mapping is also boring, I'll have a look at macros.